src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java

```diff
@@ -324,55 +318,37 @@ class ClientVector extends ArrayDeque<KeepAliveEntry> {
 
     /* return a still valid, idle HttpClient */
     HttpClient get() {
-        lock();
-        try {
-            // check the most recent connection, use if still valid
-            KeepAliveEntry e = peekFirst();
-            if (e == null) {
-                return null;
-            }
-            long currentTime = System.currentTimeMillis();
-            if ((currentTime - e.idleStartTime) > nap) {
-                return null; // all connections stale - will be cleaned up later
-            } else {
-                pollFirst();
-                if (KeepAliveCache.logger.isLoggable(PlatformLogger.Level.FINEST)) {
-                    String msg = "cached HttpClient was idle for "
-                            + Long.toString(currentTime - e.idleStartTime);
-                    KeepAliveCache.logger.finest(msg);
-                }
-                return e.hc;
+        // check the most recent connection, use if still valid
```

Comment:
It might be prudent to assert that the global cacheLock is held by the current thread in get & put (if that's feasible without too much hackery). Just worrying about future maintainers possibly modifying this code and using ClientVector outside of the lock. The assert would make sure they ponder the consequences.

Reply:
Looks reasonable. I personally dislike having several top-level classes in the same file so I'd say having ClientVector as an inner class is a plus. Having ClientVector carry a (hidden) pointer to its parent instance shouldn't be an issue since ClientVector shouldn't escape. I'd say this is good, provided that the modified test still passes. Otherwise, we can replace the assert with a simple comment, to warn that the class is not thread-safe and should not be used outside of blocks protected by the cacheLock.